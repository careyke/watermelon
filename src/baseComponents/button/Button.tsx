import React, { FC, MouseEventHandler, CSSProperties } from 'react';
import classnames from 'classnames';
import styles from './button.less';
import { ButtonType, ButtonShape, ButtonTypeClass, ButtonStyle, ButtonShapeClass, ButtonDirection, ButtonDirectionClass } from './type';

interface IButtonProps {
  type?: ButtonType,
  shape?: ButtonShape,
  icon?: string;
  buttonStyle?: ButtonStyle;
  onClick?: MouseEventHandler;
  style?: CSSProperties;
  direction?: ButtonDirection
}

export const Button: FC<IButtonProps> = (props) => {
  const { children, type = 'default', shape, buttonStyle = 'icon-text', onClick, icon, direction = 'row' } = props;
  const typeClassName = `btn${type}` as ButtonTypeClass;
  const shapeClassName = `btn${shape}` as ButtonShapeClass;
  const directionClassName = `btn${direction}` as ButtonDirectionClass;
  const buttonClass = classnames({
    [styles.button]: true,
    [styles[typeClassName]]: type,
    [styles[shapeClassName]]: shape,
    [styles[directionClassName]]: direction
  });
  const iconClass = classnames({
    [styles.icon]: true,
    [`iconfont icon-${icon}`]: icon
  });
  const textClass = classnames({
    [styles.btnicontext]: icon && buttonStyle === 'icon-text',
    [styles.btntexticon]: icon && buttonStyle === 'text-icon'
  });

  return (
    <div className={buttonClass} onClick={onClick}>
      {buttonStyle === 'icon-text' && icon ? <span className={iconClass}></span> : null}
      {shape === 'circle' && icon ? null : <span className={textClass}>{children}</span>}
      {buttonStyle === 'text-icon' && icon ? <span className={iconClass}></span> : null}
    </div>
  );
}