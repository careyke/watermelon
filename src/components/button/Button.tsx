import React, { useState } from 'react';
import classnames from 'classnames';
import { IButtonProps, ButtonType, ButtonShape, ButtonShowStyle } from './type';
import styles from './button.less';

export default function Button(props: IButtonProps): React.ReactElement {
  const { type = 'default', shape, loading = false, showStyle = 'icon-text', icon } = props;
  let iconClass = classnames({
    [styles['btn-icon']]: true,
    'iconfont': true,
    ['icon-' + icon]: true
  })
  let btnClass = classnames({
    [styles['btn-base']]: true,
    [styles['btn-default']]: type === ButtonType.Default,
    [styles['btn-primary']]: type === ButtonType.Primary,
    [styles['btn-danger']]: type === ButtonType.Danger,
    [styles['btn-link']]: type === ButtonType.Link,
    [styles['btn-round']]: shape === ButtonShape.Round,
    [styles['btn-circle']]: shape === ButtonShape.Circle,
  })
  let btnTextClass = classnames({
    [styles['btn-text']]: true,
    [styles['btn-icon-text']]: icon && showStyle === ButtonShowStyle.IconText,
    [styles['btn-text-icon']]: icon && showStyle === ButtonShowStyle.TextIcon,
  })
  // const [isLoading, setLoading] = useState(loading);

  return (
    <div className={btnClass}>
      {
        showStyle === ButtonShowStyle.IconText && icon ? <span className={iconClass} ></span> : null
      }
      {shape !== ButtonShape.Circle ? <span className={btnTextClass}>{props.children}</span> : null}
      {
        showStyle === ButtonShowStyle.TextIcon && icon ? <span className={iconClass} ></span> : null
      }
    </div>
  )
}