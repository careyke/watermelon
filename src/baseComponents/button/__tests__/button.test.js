import React from "react";
import { render, mount, shallow } from "enzyme";
import { Button } from "../Button";


describe("Button test", () => {
  it("renders correctly", () => {
    const wrapper = render(<Button>Default</Button>);
    expect(wrapper).toMatchSnapshot();
    const wrapper1 = render(<Button icon='search'>Search</Button>);
    expect(wrapper1).toMatchSnapshot();
    const wrapper2 = render(<Button type='success' icon='search'>Search</Button>);
    expect(wrapper2).toMatchSnapshot();
    const wrapper3 = render(<Button type='success' icon='search' shape='circle'>Search</Button>);
    expect(wrapper3).toMatchSnapshot();
    const wrapper4 = render(<Button type='success' icon='search' direction='col'>Search</Button>);
    expect(wrapper4).toMatchSnapshot();
  });
  it('has correct type className', () => {
    const getType = () => {
      const typeArr = ['default', 'danger', 'primary', 'warning', 'link'];
      const num = ~~(Math.random() * typeArr.length);
      return typeArr[num];
    }
    const type = getType();
    const wrapper = shallow(<Button type={type}>Default</Button>);
    expect(wrapper.hasClass(`btn${type}`)).toEqual(true);
  });
  it('has correct shape className', () => {
    const getShape = () => {
      const arr = ['round', 'circle'];
      const num = ~~(Math.random() * arr.length);
      return arr[num];
    }
    const shape = getShape();
    const wrapper = shallow(<Button shape={shape}>Default</Button>);
    expect(wrapper.hasClass(`btn${shape}`)).toEqual(true);
  });
  it('has correct direction className',()=>{
    const getDirection = () => {
      const arr = ['row', 'col'];
      const num = ~~(Math.random() * arr.length);
      return arr[num];
    }
    const direction = getDirection();
    const wrapper = shallow(<Button icon='search' direction={direction}>Default</Button>);
    expect(wrapper.hasClass(`btn${direction}`)).toEqual(true);
  });
  it('render children count correctly', () => {
    const wrapper1 = shallow(<Button icon='search'>Default</Button>);
    expect(wrapper1.children().length).toEqual(2);
    const wrapper2 = shallow(<Button>Default</Button>);
    expect(wrapper2.children().length).toEqual(1);
    const wrapper3 = shallow(<Button shape='circle' icon='search'>Default</Button>);
    expect(wrapper3.children().length).toEqual(1);
  });
  it('render icon order correctly', () => {
    const wrapper1 = shallow(<Button icon='search'>Search</Button>);
    expect(wrapper1.childAt(0).hasClass('icon')).toEqual(true);
    const wrapper2 = shallow(<Button buttonStyle='text-icon' icon='search'>Search</Button>);
    expect(wrapper2.children().last().hasClass('icon')).toEqual(true);
  });
  it('trigger click correctly', () => {
    const handleClick = jest.fn(); //Returns a new, unused mock function,expect() expect it;
    const wrapper = mount(<Button onClick={handleClick}>Click</Button>);
    wrapper.simulate('click');
    expect(handleClick).toHaveBeenCalled();
  });
  it("render className props correctly", () => {
    const wrapper = shallow(<Button className='testClassName' />);
    expect(wrapper.hasClass('testClassName')).toEqual(true);
  });
});
