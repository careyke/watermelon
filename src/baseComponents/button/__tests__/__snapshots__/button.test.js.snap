// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button test renders correctly 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button btndefault btnrow",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "",
        },
        "children": Array [
          Object {
            "data": "Default",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly 2`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button btndefault btnrow",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "btnicontext",
          },
          "children": Array [
            Object {
              "data": "Search",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "btnicontext",
        },
        "children": Array [
          Object {
            "data": "Search",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "icon iconfont icon-search",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly 3`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button btnsuccess btnrow",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "btnicontext",
          },
          "children": Array [
            Object {
              "data": "Search",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "btnicontext",
        },
        "children": Array [
          Object {
            "data": "Search",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "icon iconfont icon-search",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly 4`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button btnsuccess btncircle btnrow",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;

exports[`Button test renders correctly 5`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "button btnsuccess btncol",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "icon iconfont icon-search",
        },
        "children": Array [],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "btnicontext",
          },
          "children": Array [
            Object {
              "data": "Search",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "btnicontext",
        },
        "children": Array [
          Object {
            "data": "Search",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "span",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "icon iconfont icon-search",
          },
          "children": Array [],
          "name": "span",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
